Add M-Vave SMC-Mixer controller mapping

SamWhited opened the pull request:
This PR adds support for the M-Vave SMC-Mixer.
The corresponding documentation PR is here: https://github.com/mixxxdj/manual/pull/727

I have gone ahead and signed the CLA for this one.

SamWhited committed:
Add M-Vave SMC-Mixer controller mapping

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Rebased and changed the bsae branch to 2.5, oops!

SamWhited committed:
Fix misspelling in M-Vave SMC-Mixer XML file

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Pushed up a new commit fixing the misspelling detected by CI

Swiftb0y commented:
Ah yes. A ComponentsJS mapping. A man of culture. I will take a look asap. Feel free to remind me in case I forget.

SamWhited committed:
Apply linter patch to M-Vave SMC-Mixer scripts

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Pushed up a new commit applying the patch recommended by pre-commit.

SamWhited committed:
Remove unused function in SMC-Mixer script

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Apply a few linter fixes

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
More fixes suggested by eslint

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
It seems that eslint replaces the var for the exported namespace with a "let", breaking the mapping. If I try to add the "eslint-ignore-next-line" comment I see in some other files in the repo it doesn't seem to work and locally it still gets replaced. Suggestions welcome.

SamWhited committed:
Remove some unnecessary exports

Signed-off-by: Sam Whited <sam@samwhited.com>

Swiftb0y commented:
> Applied the patch suggested by the linter for real this time. Unsure why if I run it locally it doesn't complain about this. Hopefully CI will be happy now!

The CI runs on the entire PR diff. pre-commit locally just runs on the commit. Thats usually an issue when people only set up pre-commit after they've already made some commits to their branch. In that case, you can run the same command as CI once and apply the suggestions in a separate commit: `pre-commit run --from-ref 2.5 --to-ref HEAD`.

Swiftb0y commented:
> It seems that eslint replaces the var for the exported namespace with a "let", breaking the mapping. If I try to add the "eslint-ignore-next-line" comment I see in some other files in the repo it doesn't seem to work and locally it still gets replaced. Suggestions welcome.

I think the problem is that eslint has slightly changed their rules since then and `no-unused-vars` is the wrong rule (thus eslint replaces it and breaks the mapping). I think you want to disable `prefer-const` or `no-var`

Swiftb0y commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -0,0 +1,469 @@
+                }
+            };
+            this.inToggle = function() {
+                if (this.isLongPressed) {
+                    if (this.knob.key === this.key.replace("button_", "") || (this.key === "enabled" && this.knob.key === "super1")) {
+                        this.knob.group = this.origGroup;
+                        this.knob.key = this.origKey;
+                        this.knob.inKey = this.origInKey;
+                        this.knob.outKey = this.origOutKey;
+                    } else {
+                        this.knob.group = this.group;
+                        let newKey = "";
+                        if (this.key === "enabled") {
+                            newKey = "super1";
+                        } else {
+                            newKey = this.key.replace("button_", "");
+                        }
+                        this.knob.key = newKey;
+                        this.knob.inKey = newKey;
```
can you explain some of the special sauce you do here? Why is it necessary? What are you trying to accomplish?

SamWhited replied:
I'll add a comment to make it clearer. The general gist is that there's only one rotary encoder per knob at the top, and we want to be able to use it for all the EQ buttons, but also have cutoffs. Short pressing the button does the cutoff, and long pressing changes the knob to be that EQ control. If we use the regular power window the short press always fires, so the cutoff is turned on and you hear it even though that's not what you meant to do.

SamWhited replied:
Oh, or you may have meant all the "button" and "super1" stuff. That's to set the knob to change the value affected by the button (ie. if we hit `button_parameter1` we should have the knob's key set to `parameter1`, or for `enabled` we want it set to `super1`, the quick effects button). I agree this is confusing. I'll see if I can't think of a way to simplify it, maybe just with a big switch statement that would be easier to read, or add a comment if not.

Swiftb0y replied:
So the pot changes behavior depending on the button state? Then there are two established ways to deal with that: 
1. Instance two separate components and then route the `input` where you want it to go. You can do that by simply having both as local variable and then assign the instance you want to dispatch to in the button handler. 
2. You look at how the `shift` and `unshift` stuff is done. Each component has a function that reconfigures it on the fly. Then you would do a `c.disconnect(); c.shift(); c.connect(); c.trigger();` (where `c` is the component instance you need to reconfigure). In your case, you would set new `inKey`s and `group`s. 

Does that make sense?

Swiftb0y replied:
Its primary a codestyle question. The class defines the behavior and the instance supplies the data. In your current solution, the both are being conflated because the class makes explicit assumptions on what values (COs) its being used with. So I don't consider the current solution particular clean.

SamWhited replied:
Oh right, I see what you're saying; I think I had it the "correct" way originally, then got lazy and decided the class was only used in this one file anyways so I changed it so that I wouldn't have to assign the method to every instance.

I'm generally a fan of not pre-generalizing something that will never be used in any other file just for the sake of technically being more robust to future use, but I'll go try to swap it around and see if I can make it work in a clean way.

Swiftb0y replied:
yeah, definitely good guidance, but in this case the generalization would result in stricter separation of concerns and thus increased code clarity. I wouldn't worry too much about duplication of the instances right now, its very easy to combat that (use loops and LUTs, or merge partial Object definitions using `Object.assign`).

SamWhited replied:
I moved this back to where each instance is defined and curried in the various values it needs to store so that I only had to define the function in one place. Hopefully that's a little cleaner; LMK what you think, and thanks for all your help with this review!

Swiftb0y reviewed (commented):
This is in a very good shape already

SamWhited committed:
Use inValueScale in SMC-Mixer scripts

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

SamWhited commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -0,0 +1,461 @@
+                }
+            };
+        }
+    }
```
If I commit this it complains that the JavaScript isn't valid:

> 37:8  error  Parsing error: Unexpected token. A constructor, method, accessor, or property was expected

But, not really knowing any JavaScript, it's unclear to me why that would be. This seems like a pretty normal class definition that I see elsewhere in other files, and this looks like a method to me?

Swiftb0y replied:
Ah yeah sorry. Little rusty on the JS myself

SamWhited replied:
I had to revert the logic back on this one, I wasn't paying enough attention and the way it was setup results in incorrect values. I did continue to use inValueScale though and everything seems to work now.

SamWhited commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -0,0 +1,461 @@
+                },
+            });
+            // Faders
+            this.faders = new Array(9);
```
Oops, good catch, thank you.

SamWhited committed:
Fix fader array length in SMC-Mixer script

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

SamWhited commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -0,0 +1,461 @@
+            this.leftButton = new components.Button({
+                group: "[Library]",
+                midi: [0x90, 0x62],
+                key: "focused_widget",
```
I'll change it to inKey, as far as I can tell the controller handles the LED for this button (at least in the firmware I have) and there's no way to set it. They indicated to me in an email that they would have better LED control in a future firmware, but for this in particular you're looking at the library on the computer anyways, so I'm not sure that it's important to indicate where the focus is on the controller.

SamWhited commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -0,0 +1,461 @@
+            });
+            // Faders
+            this.faders = new Array(9);
+            this.faders[0] = new Pot({
+                group: "[Channel3]",
+                midi: [0xE0],
+                key: "volume",
+                softTakeover: true,
+            });
+            this.faders[1] = new Pot({
+                group: "[Channel1]",
+                midi: [0xE1],
+                key: "volume",
+                softTakeover: true,
+            });
+            this.faders[2] = new Pot({
+                group: "[Channel2]",
+                midi: [0xE2],
+                key: "volume",
+                softTakeover: true,
+            });
+            this.faders[3] = new Pot({
+                group: "[Channel4]",
+                midi: [0xE3],
+                key: "volume",
+                softTakeover: true,
+            });
```
Done, thank you!

SamWhited commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -0,0 +1,461 @@
+        constructor(params) {
+            super(params);
+            this.toleranceWindow = 0.001;
+            // TODO: why do I have to override this?
+            this.connect = function() {
+                if (undefined !== this.group
+                    && undefined !== this.outKey
+                    && undefined !== this.output
+                    && typeof this.output === "function") {
+                    this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+                }
+            };
```
No, sorry, the mapping actually fails and complains this function doesn't exist if I don't override it. I can't for the life of me see how that's possible though since it should be inherited. Let me try again, I've made a number of changes since then so maybe I had done something dumb to break however JavaScript does its inheritance and it will work now.

SamWhited replied:
Yup, this appears to work now even without overriding this function. I have no idea what was happening before. Removing it; thanks!

SamWhited commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -0,0 +1,469 @@
+    }
+    class EqRack {
+        constructor(deck) {
+            // Normalize the deck number to be zero indexed (the third deck is the 0th
+            // control).
+            let normDeck = deck;
+            switch (deck) {
+            case 3: {
+                normDeck = 0;
+                break;
+            }
+            case 4: {
+                normDeck = 3;
+                break;
+            }
```
Done.

SamWhited commented on res/controllers/MVave-SMC-Mixer-scripts.js:
```diff
@@ -0,0 +1,469 @@
+            this.origOutKey = knob.outKey;
+        }
+    }
+    class Pot extends components.Pot {
+        constructor(params) {
```
It's really hard to get the values aligned exactly, so we have a window that's considered the same value for the sake of not having the soft takeover indicator LED blinking constantly because the slider is 0.001 away from what the software expects. I think I pulled this from the Traktor S4 Mk3 mapping and it made the indicator much more useful.

SamWhited replied:
Added a comment.

Swiftb0y replied:
I'm afraid I can't follow you. What softtakeover indicator LED?

SamWhited replied:
Above each fader is an LED that blinks if the hardware fader position does not match the value in Mixxx, this way you know that soft takeover is enabled and you need to match the fader up with what Mixxx has before you can change the volume or whatever. The purpose of this class is to blink the LED when the value Mixxx sees does not match the last recorded value that was set by moving the fader.

SamWhited committed:
Lots of minor review changes to SMC-Mixer scripts

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Fix EQ buttons broken by review patch

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Fix no-unused-vars pragma in SMC-Mixer script

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
More pre-commit fixes

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Add comment about why Pot class is necessary

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Add USB hardware ID to XML file

I'm unsure if this actually does anything for MIDI devices, but it may
be good to have just for documentation purposes.

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Fixed logic for encoders, broken during review

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited committed:
Move instance specific behavior out of class

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
I also reverted adding the hardware ID to the XML file. It appears they use the same one for multiple devices (probably it's the same MIDI chip internally), so it's not actually useful for detecting what device this is.

SamWhited committed:
Rotary encoders don't have LEDs, remove outKey

Signed-off-by: Sam Whited <sam@samwhited.com>

Swiftb0y commented:
Thats good to know. The hardware ID is not used for anything atm anyways (the underlying APIs are so generic they don't give us access to that) and only has documentary value. So in that case it would be nice if the data was present and you just add an XML comment that documents that.

SamWhited committed:
Add comment about vendor/product ID to XML file

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Gotcha; I've added it back and added a comment to let people know that it's not unique.

SamWhited commented:
Finally green! Thanks for all your help and for kicking off CI so many times.

Swiftb0y commented:
No worries. I'll do a second review tomorrow.

Swiftb0y reviewed (approved):
No obvious objections. This is good to merge. I'll take a look at the manual.
